### PR TITLE
Improve tweakreg resilience to empty catalogs

### PR DESCRIPTION
**Pull request by mcara**
This PR fixes https://github.com/spacetelescope/jwst/issues/4261

### CODE REVIEW

File: jwst/tweakreg/tweakreg_step.py [hunk 1]
```diff
@@ -170,17 +170,35 @@ def process(self, input):
...
+                "non-empty catalogs."):
+                # we need at least two exposures to perform image alignment
+                self.log.warning(msg)
+                self.log.warning("At least two exposures are required for "
+                                 "image alignment.")
+                self.log.warning("Nothing to do. Skipping 'TweakRegStep'...")
+                self.skip = True
+                for model in images:
+                    model.meta.cal_step.tweakreg = "SKIPPED"
+                return input
```
Comment (jdavies-st) [inline jwst/tweakreg/tweakreg_step.py]: This should probably return `images`, right?

Comment (mcara) [reply jwst/tweakreg/tweakreg_step.py]: No, `input` is perfect.

Comment (mcara) [reply jwst/tweakreg/tweakreg_step.py]: Yes, you are right.

File: jwst/tweakreg/tweakreg_step.py [hunk 2]
```diff
@@ -170,17 +170,35 @@ def process(self, input):
...
+                sigma=(self.sigma, 'rmse')
+            )
+
+        except ValueError as e:
+            msg = e.args[0]
+            if (msg == "Too few input images (or groups of images) with "
+                "non-empty catalogs."):
+                # we need at least two exposures to perform image alignment
+                self.log.warning(msg)
+                self.log.warning("At least two exposures are required for "
```
Comment (hbushouse) [inline jwst/tweakreg/tweakreg_step.py]: Too late now, but in hind sight I think this warning message could've been a bit more explicit. The problem condition that's being encountered is not the fact that there aren't enough images, it's the fact that there aren't enough images **WITH DETECTED SOURCES**. I think we'll get confused users saying "I have several input images, so why is it telling me I need at least two?".

Comment (jdavies-st) [reply jwst/tweakreg/tweakreg_step.py]: Is the message given above, "Too few input images (or groups of images) with non-empty catalogs" not clear enough?  Or is the term "detected sources" preferable?

Comment (hbushouse) [reply jwst/tweakreg/tweakreg_step.py]: If the message referred to in lines 188-189 is also seen by the user, then I think we're OK as-is. I had initially assumed that that message was being trapped and then only the one in lines 192-193 was being written out for the user.

**Comment by hbushouse** [comment]
Just tested this fix on the example data in /grp/jwst/ssb/chartreuse/dms_bugs/jwstdms-245/ and it still fails. The new try/except block around the call to align_wcs() is correctly branching to the except part of the block, but it's looking for the wrong error message. The error message being returned is "Image catalog must contain at least one entry.". Because it's not finding this error message, the remaining part of the except block is not executed and the step still fails with a raised exception.

**Comment by hbushouse** [comment]
When I modify the code to look for the appropriate error message, it then correctly skips the rest of the ``tweakreg`` step, but the only warnings sent to the user are:
```
2019-11-25 11:13:56,656 - stpipe.Image3Pipeline.tweakreg - WARNING - Image catalog must contain at least one entry.
2019-11-25 11:13:56,656 - stpipe.Image3Pipeline.tweakreg - WARNING - At least two exposures are required for image alignment.
2019-11-25 11:13:56,656 - stpipe.Image3Pipeline.tweakreg - WARNING - Nothing to do. Skipping 'TweakRegStep'...
```
So nowhere does it say anything about not having any sources in any of the images.

**Comment by philhodge** [comment]
Testing on an error message seems particularly fragile to me.  In order to check for a particular error, if you don't want to bother setting up different classes, I would suggest assigning integer codes and testing for the code.

**Comment by mcara** [comment]
> if you don't want to bother setting up different classes

I do not see the difference between testing on an integer or a bunch of characters (`int8`). Same thing to me. I thought about introducing a new exception class but it really does not make much sense from the point of view of the stand-alone `tweakwcs`.

**Comment by jdavies-st** [comment]
Yes, agree that excepting on an error message is fragile.  But don't introduce integer error codes either.   The way to to this in python is to use an Exception class.  And if `tweakwcs` gives errors of the same class that needed differentiation by the wrapping code in `jwst.tweakreg`, then use a custom error class.

```python
try:
    stuff
except (ValueError, MyException):
    other stuff
```

And `MyException` should probably subclass `ValueError` instead of `Exception`.  Though maybe this is overkill, and `ValueError` is enough without the `if` block.

**Comment by mcara** [comment]
@hbushouse 
>  The error message being returned is "Image catalog must contain at least one entry.". Because it's not finding this error message, the remaining part of the except block is not executed and the step still fails with a raised exception.

This is a very different message from what @jdavies-st and I were getting. Please make sure you are using correct environment with an updated `tweakwcs`. Something is not right here.

**Comment by hbushouse** [comment]
You're absolutely right! I hadn't updated the rest of my environment in a while and was obviously running an older version of `tweakwcs`. After creating a new environment I now get:
```
2019-11-25 13:24:44,694 - stpipe.Image3Pipeline.tweakreg - WARNING - Group with ID 'jw87700005001_02101_00001_nis_cal' will not be aligned: empty source catalog
2019-11-25 13:24:44,698 - stpipe.Image3Pipeline.tweakreg - WARNING - Group with ID 'jw87700005001_02103_00001_nis_cal' will not be aligned: empty source catalog
2019-11-25 13:24:44,701 - stpipe.Image3Pipeline.tweakreg - WARNING - Group with ID 'jw87700005001_02104_00001_nis_cal' will not be aligned: empty source catalog
2019-11-25 13:24:44,708 - stpipe.Image3Pipeline.tweakreg - WARNING - Too few input images (or groups of images) with non-empty catalogs.
2019-11-25 13:24:44,708 - stpipe.Image3Pipeline.tweakreg - WARNING - At least two exposures are required for image alignment.
2019-11-25 13:24:44,708 - stpipe.Image3Pipeline.tweakreg - WARNING - Nothing to do. Skipping 'TweakRegStep'...
2019-11-25 13:24:44,711 - stpipe.Image3Pipeline.tweakreg - INFO - Step tweakreg done
2019-11-25 13:24:44,711 - stpipe.Image3Pipeline - INFO - Matching sky values across all input images...
```
just as it's supposed to be.

Sorry for the spam. Will close the ticket again.

**Pull request merged**